Source/Concurrency/CivlTypeChecker.cs

```diff
@@ -324,7 +324,20 @@ private void TypeCheckActions()
       {
         Error(program, "Call graph over atomic actions must be acyclic");
       }
-
+      
+      actionProcs.Iter(proc =>
+      {
+        if (proc.FindAttribute("inline") != null)
+        {
+          Error(proc, "unnecessary to provide inline attribute on action");
+        }
+        var impl = actionProcToImpl[proc];
+        if (impl.FindAttribute("inline") != null)
+        {
+          Error(impl, "unnecessary to provide inline attribute on action");
+        }
+      });
```

Comment:
I know that in Boogie procedures and implementations can be declared separately, but that if declared together (i.e., as a procedure with body) the set of supplied attributes is given to both. What I'm wondering is: does Boogie processes the inline attribute on both, or if it is meaningful on only one of them? If it is the latter, we could only do the corresponding check and avoid duplicate error messages.

Reply:
```
protected int GetInlineCount(Implementation impl)
    {
      Contract.Requires(impl != null);
      Contract.Requires(impl.Proc != null);

      string /*!*/
        procName = impl.Name;
      Contract.Assert(procName != null);
      if (recursiveProcUnrollMap.TryGetValue(procName, out var c))
      {
        return c;
      }

      c = -1; // TryGetValue above always overwrites c
      impl.CheckIntAttribute("inline", ref c);
      // procedure attribute overrides implementation
      impl.Proc.CheckIntAttribute("inline", ref c);

      recursiveProcUnrollMap[procName] = c;
      return c;
    }
```

The code above in Inline.cs examines the `inline` attribute. It seems that the attribute on the procedure overrides the one on the implementation. So I think I can avoid both the check and the injection of the inline attribute on the implementation.